### PR TITLE
Bug fixes

### LINKED ISSUE #3582: coursera jhu swirl library completion token paste not working
Please briefly describe what you were doing that led to the issue, if applicable:

The coursera john hopkins university data science specialty [ see aka.ms/dsJhu ] classes make use of a library called swirl to walk students through exercises. when you get to the end of any exercise it asks you if you want to submit completion of the exercise at which point it asks for email address and token the latter of which you copy and paste from your course web site view for this exercise.   
In rstudio you can copy that token from browser page and paste into R interactive window at that point in swirl library routine w/o any issue.  Using vs17/dev15.2 rtools R interactive window it doesn't for whatever reason allow that token, or any alphanumeric value, paste from clipboard to happen.  It does accept you typing input which is current work around that requires you to bring up view of lengthy token and manually try and type it into that submission prompt for token.

The simplest way i can think of to repro this is to kick off one of the simple swirl exercises using the following commands and stepping through the <10min easy q&a exercise, easy that is for an experienced R programmer, at which point you can repro it not allowing you to paste any value from clipboard into R interactive window when you get to it asking you for completion submission token.

specific repro steps involve doing the following from R interactive command line
if (!require("swirl")) { install.packages("swirl"); packageVersion("swirl"); library(swirl) }
install_from_swirl("R Programming"); swirl()
# select "R Programming" option | <any option, e.g. 1. Basic Building Blocks> |
# step through <10min q&a exercises till get to point where they ask you to submit results
# at which point enter an email address and when it next ask for a token try and paste any 
# alphanumeric value from your clipboard into the line. with rstudio this works but with 
# vs17/dev15.2 rtools this doesn't accept paste from clipboard.  it does accept you typing input.

If you have any screenshots demonstrating the issue, please include them as well to help us diagnose it better.

Additional information:

OS Information
    Version:       Microsoft Windows NT 10.0.15063.0

RTVS Information:
    Assembly: Microsoft.VisualStudio.R.Package, Version=1.1.30523.1233, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a

Active R URI:
    R 3.3.2: C:\Program Files\R\R-3.3.2

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#3582 coursera jhu swirl library completion token paste not working
#3655 Uses of 'T'/'F' incorrectly flagged as syntax errors
#3656 Roxygen comments for S4 function not added
Fix template deployment on F5

### CODE REVIEW

File: src/Package/Impl/Sql/ConnectionStringEditor.cs [hunk 1]
```diff
@@ -41,7 +41,7 @@ internal sealed class ConnectionStringEditor : UITypeEditor {
         /// </summary>
         [Export(typeof(IConfigurationSettingUIEditorProvider))]
         [Name(ConnectionStringEditor.ConnectionStringEditorName)]
-        class EditorProvider : IConfigurationSettingUIEditorProvider {
+        private class EditorProvider : IConfigurationSettingUIEditorProvider {
```
Comment (AlexanderSher) [inline src/Package/Impl/Sql/ConnectionStringEditor.cs]: Does `ExportAttribute` work on private classes?

File: src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs [hunk 2]
```diff
@@ -31,13 +31,8 @@ public static class AstRootExtensions {
...
             if (func != null) {
                 return func;
-            } else {
-                var v = scope.FindVariableDefinitionByName(itemName, position);
-                if (v != null) {
-                    return v;
-                }
             }
-            return null;
+            return scope.FindVariableDefinitionByName(itemName, position);
```
Comment (AlexanderSher) [inline src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs]: `return func ?? scope.FindVariableDefinitionByName(itemName, position);`

File: src/R/Core/Impl/Tokens/RTokenizer.cs [hunk 3]
```diff
@@ -332,6 +326,36 @@ public sealed class RTokenizer : BaseTokenizer<RToken> {
...
         }
 
+        /// <summary>
+        /// Tries to determine if T/F are actually logical values.
+        /// </summary>
+        /// <remarks>This is not 100% reliable during tokenization since T or F may be 
+        /// part of expression such as T &lt;- 0 and then, 100 lines later, `x = T + 2`. 
+        /// However, in the latter case we can leave T/F as logical and let expression
+        /// parser figure out the context.</remarks>
+        private bool VerifyLogical(string s) {
```
Comment (AlexanderSher) [inline src/R/Core/Impl/Tokens/RTokenizer.cs]: `IsLogical`?

File: src/R/Core/Test/Tokens/TokenizeIdentifierTest.cs [hunk 4]
```diff
@@ -63,15 +64,15 @@ public class TokenizeIdentifierTest : TokenizeTestBase<RToken, RTokenType> {
...
-            var tokens = Tokenize("1 <- F", new RTokenizer());
-
-            tokens.Should().Equal(new [] {
-                RTokenType.Number,
-                RTokenType.Operator,
-                RTokenType.Logical,
-            }, (token, tokenType) => token.TokenType == tokenType);
+        [CompositeTest]
+        [InlineData("1 <- F", 2, RTokenType.Logical)]
+        [InlineData("F <- 1", 0, RTokenType.Identifier)]
```
Comment (AlexanderSher) [inline src/R/Core/Test/Tokens/TokenizeIdentifierTest.cs]: Do you need any invalid code to be tested, something like "F =<- 1" of "F))"?

File: src/R/Editor/Impl/Roxygen/RoxygenBlock.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+using Microsoft.R.Core.AST;
+using Microsoft.R.Core.AST.Arguments;
+using Microsoft.R.Core.AST.Extensions;
+using Microsoft.R.Core.AST.Functions;
+using Microsoft.R.Core.AST.Operators;
+using Microsoft.R.Core.AST.Variables;
+using static System.FormattableString;
+
+namespace Microsoft.R.Editor.Roxygen {
+    public static class RoxygenBlock {
```
Comment (AlexanderSher) [inline src/R/Editor/Impl/Roxygen/RoxygenBlock.cs]: Are there any changes in this file?

File: src/R/Editor/Impl/Tree/TextChangeAnalyzer.cs [hunk 6]
```diff
@@ -64,6 +68,34 @@ internal static class TextChangeAnalyzer {
...
+            }
+
+            // Check if line break is added to removed on a line with closing } of 'if'
+            // so we can full-parse if 'else' position changes relatively to the if
+            var snapshot = context.EditorTree.EditorBuffer.CurrentSnapshot;
+            var text = snapshot.GetLineFromPosition(context.PendingChanges.Start).GetText();
+
+            // We need to find if any position in the line belong to `if` scope
+            // if there is 'else' the same line
+            return FindKeyword("if", text) || FindKeyword("else", text);
```
Comment (AlexanderSher) [inline src/R/Editor/Impl/Tree/TextChangeAnalyzer.cs]: `RTokenizer` can be reused (e.g. `FindAnyKeyword(text, "if", "else")`). Does it matter in terms of performance?

File: src/R/Core/Test/Tokens/TokenizeIdentifierTest.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Core/Test/Tokens/TokenizeIdentifierTest.cs]: This is specifically to tread T/F as identifier in certain expressions, but sure.

File: src/R/Editor/Impl/Roxygen/RoxygenBlock.cs (hunk 5 above)
Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Roxygen/RoxygenBlock.cs]: Yes, added support for S4. Looks like when file moves git loses diff tracking...

File: src/R/Editor/Impl/Tree/TextChangeAnalyzer.cs (hunk 6 above)
Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Tree/TextChangeAnalyzer.cs]: Constructor is tiny so prob not worth it.

**Comment by MikhailArkhipov** [comment]
Can't see some of comments, (CF?).

`VerifyLogical` - there is `Logicals.IsLogical(s)` check up above so didn't want to cause confusion.
Yes, export seems to work on private classes

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**